avocado/plugins/resolvers.py

```diff
@@ -110,6 +110,25 @@ def resolve(reference):
                                find_avocado_tests)
 
 
+class RequirementsResolver:
+
+    name = 'requirements'
+    description = 'Requirements resolver for tests with requirements'
+
+    @staticmethod
+    def resolve(runnable):
+        requirements_runnables = []
+        for requirement in runnable.requirements:
+            requirement_copy = requirement.copy()
+            kind = 'requirement-%s' % requirement_copy.pop('type')
+            args = requirement_copy.values()
```

Comment:
Right, this makes things easy on the runner side. We don't need the `type` inside the runner, but when we have other kinds of runners, like one that handles assets, we will have more arguments that can be passed to the runner without handling it here.

In a nutshell, my idea is that the resolver's code should be abstract enough to pass the arguments to the runner so that they can be used right away.

Reply:
> Right, this makes things easy on the runner side. We don't need the `type` inside the runner, but when we have other kinds of runners, like one that handles assets, we will have more arguments that can be passed to the runner without handling it here.
> 
> In a nutshell, my idea is that the resolver's code should be abstract enough to pass the arguments to the runner so that they can be used right away.

I don't think it's good idea to get requirements passed by name (from the safeloader's parsing of the docstring) and turn them into a list.  This is too fragile, like in the example given above.

If you keep the code you're proposing, and incorporate your idea bout supporting listing, removing, you'd allow users to write:

```
:avocado: requirement={"type": "package", "name": "vim-common", "action": "install"}
:avocado: requirement={"type": "package", "action": "remove", "package": "emacs"}
```

And end up with and attempt to either:

1.1) install the vim-common (package) 
1.2) "emacs" a "remove" (non-sense)
2.1) vim-common an install (non-sense)
2.2) remove the emacs (package)

Reply:
> > Right, this makes things easy on the runner side. We don't need the `type` inside the runner, but when we have other kinds of runners, like one that handles assets, we will have more arguments that can be passed to the runner without handling it here.
> > In a nutshell, my idea is that the resolver's code should be abstract enough to pass the arguments to the runner so that they can be used right away.
> 
> I don't think it's good idea to get requirements passed by name (from the safeloader's parsing of the docstring) and turn them into a list. This is too fragile, like in the example given above.
> 

You are right, they should be handled as kwargs in the runner. Your example made me see that.

> If you keep the code you're proposing, and incorporate your idea bout supporting listing, removing, you'd allow users to write:
> 
> ```
> :avocado: requirement={"type": "package", "name": "vim-common", "action": "install"}
> :avocado: requirement={"type": "package", "action": "remove", "package": "emacs"}
> ```
> 
> And end up with and attempt to either:
> 
> 1.1) install the vim-common (package)
> 1.2) "emacs" a "remove" (non-sense)
> 2.1) vim-common an install (non-sense)
> 2.2) remove the emacs (package)